jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/queue/QueueStatistics.kt

```diff
@@ -0,0 +1,227 @@
+import java.util.IdentityHashMap
+import java.util.concurrent.ConcurrentHashMap
+
+class QueueStatisticsInternal(queueSize: Int) {
+    /**
+     * Total packets added to the queue.
+     */
+    private val totalPacketsAdded = LongAdder()
+
+    /**
+     * Total packets removed to the queue.
+     */
+    private val totalPacketsRemoved = LongAdder()
+
+    /**
+     * Total packets dropped from the queue.
+     */
+    private val totalPacketsDropped = LongAdder()
+
+    /**
+     * The time the first packet was added.
+     */
+    private var firstPacketAddedMs: Long = -1
+
+    /**
+     * Statistics about queue lengths
+     */
+    private val queueLengthStats = BucketStats(lengthBucketSizes(queueSize), "_queue_size_at_remove", "")
+
+    /**
+     * Statistics about the time that packets were waiting in the queue.
+     */
+    private val queueWaitStats = if (QueueStatistics.trackTimes) BucketStats(waitBucketSizes, "_queue_wait_time_ms", " ms") else null
+
+    /**
+     * Gets a snapshot of the stats in JSON format.
+     */
+    val stats: OrderedJsonObject
+        get() {
+            val stats = OrderedJsonObject()
+            val now = System.currentTimeMillis()
+            stats["added"] = totalPacketsAdded.sum()
+            stats["removed"] = totalPacketsRemoved.sum()
+            stats["dropped"] = totalPacketsDropped.sum()
+            val duration = (now - firstPacketAddedMs) / 1000.0
+            stats["duration_s"] = duration
+            val packetsRemoved = totalPacketsRemoved.sum().toDouble()
+            stats["average_remove_rate_pps"] = packetsRemoved / duration
+            stats["queue_size_at_remove"] = queueLengthStats.toJson()
+            queueWaitStats?.let { stats["queue_wait_time"] = it.toJson() }
+            return stats
+        }
+
+    /**
+     * Registers the addition of a packet.
+     */
+    fun added(now: Long) {
+        if (firstPacketAddedMs < 0) {
+            firstPacketAddedMs = now
+        }
+        totalPacketsAdded.increment()
+    }
+
+    /**
+     * Registers the removal of a packet.
+     */
+    fun removed(now: Long, queueSize: Int, waitTime: Long?) {
+        totalPacketsRemoved.increment()
+        queueLengthStats.addValue(queueSize.toLong())
+        if (waitTime != null) {
+            queueWaitStats?.addValue(waitTime)
+        }
+    }
+
+    /**
+     * Registers that a packet was dropped.
+     */
+    fun dropped(now: Long) {
+        totalPacketsDropped.increment()
+    }
+
+    companion object {
+        /**
+         * The scale to use for [RateStatistics]. 1000 means units
+         * (e.g. packets) per second.
+         */
+        private const val SCALE = 1000
+
+        /**
+         * The interval for which to calculate rates in milliseconds.
+         */
+        private const val INTERVAL_MS = 5000
+
+        /**
+         * Calculate the statistics buckets for a given queue length
+         */
+        private fun lengthBucketSizes(size: Int): LongArray {
```

Comment:
nit: `size` -> `queueLength` (or `queueSize`), since that's what it seems to represent?

Reply:
nit: `lengthBucketSizes` -> `getLengthBucketSizes` ?